lib/shared/bucketing-assembly-script/assembly/eventQueue/requestPayloadManager.ts

```diff
@@ -0,0 +1,173 @@
+
+                            const metaData = new JSON.Obj()
+                            metaData.set('_feature', _feature)
+                            metaData.set('_variation', _variation)
+
+                            // Add aggVariableEvaluated Events
+                            const dvcEvent = new DVCEvent(type, variableKey, null, value, metaData)
+                            aggEvents.push(new DVCRequestEvent(dvcEvent, user_id, null))
+                        }
+                    }
+                }
+            }
+        }
+
+        // Generate defaulted agreagate user as our events APIs require a user / user_id
+        const dvcUser = new DVCPopulatedUser(new DVCUser(
+            user_id, null, null, null, null, NaN, null, null, null, null
+        ))
+        this.addEventsToPendingPayloads(new UserEventsBatchRecord(dvcUser, aggEvents))
+    }
+
+    /**
+     * Chunk up UserEventsBatchRecord's into payloads of size: this.chunkSize
+     */
+    private addEventsToPendingPayloads(record: UserEventsBatchRecord): void {
+        let flushPayload = new FlushPayload([])
+
+        while (record.events.length > this.chunkSize) {
+            const batchRecord = new UserEventsBatchRecord(
+                record.user,
+                record.events.splice(0, this.chunkSize)
+            )
+            flushPayload.records.push(batchRecord)
+            this.pendingPayloads.set(flushPayload.payloadId, flushPayload)
+            flushPayload = new FlushPayload([])
+        }
+        if (record.events.length > 0) {
+            flushPayload.records.push(record)
+        }
+        if (flushPayload.records.length > 0) {
+            this.pendingPayloads.set(flushPayload.payloadId, flushPayload)
+        }
+    }
+
+    /**
+     * Mark pending payload as success, remove from pending payloads array.
+     */
+    markPayloadSuccess(payloadId: string): void {
+        if (!this.pendingPayloads.has(payloadId)) {
+            throw new Error(`Could not find payloadId: ${payloadId} to mark as success`)
+        }
+
+        this.pendingPayloads.delete(payloadId)
+    }
+
+    markPayloadFailure(payloadId: string, retryable: boolean): void {
+        if (!this.pendingPayloads.has(payloadId)) {
+            throw new Error(
+                `Could not find payload: ${payloadId}, retryable: ${retryable} to mark as failure`
+            )
+        }
+
+        if (retryable) {
+            const payload = this.pendingPayloads.get(payloadId)
+            payload.status = 'failed'
+        } else {
+            this.pendingPayloads.delete(payloadId)
+        }
+    }
+
+    /**
+     * Check that the pendingPayloads queue is empty or only contains retryable failed payloads
+     * before creating new batch of payloads.
+     */
+    checkForFailedPayloads(): void {
+        this.pendingPayloads.values().forEach((payload) => {
+            if (payload.status === 'failed') {
+                payload.status = 'sending'
```

Comment:
won't this status change be incorrect if the code encounters a failed payload first, but then ends up throwing an error due to a different payload in the queue later on? Then you'll have changed some of the failed payloads to "sending" before throwing the error.